src/Reflection/Visibility.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Reflection;
+
+/**
+ * @internal
+ */
+final class Visibility
```

Comment:
For me this class smells the "I want a real enum in PHP" even at the cost of working around the language too much. IMO just having a `public const ALL = [self::PUBLIC, self::PROTECTED]` instead of L46-L74 would be enough. You can add an `Assert::oneOf($visibility, Visibility::ALL)` statement if you wish.

Reply:
Its probably a bit overkill, but now this class has an API that can't be worngly used.

Reply:
True, but now this class can't be used wrong, but i agree its a bit overkill

Reply:
I think it's the kind of decision that fits better in a dedicated PR with an ADR entry tbh: it's not critical to the PR and should be relatively easy to enforce everywhere in one go

On a side note: if you wanted a close real enum, it would be preferable to implement the singleton pattern there to ease the comparison of two instances

Reply:
Still don't like it, but I'll leave it to the others to whether they are fine with it or would prefer the issue to be tackled in a dedicated PR